### PR TITLE
Kata Container support

### PR DESCRIPTION
**Pull request by pohly**
Contains support for creating image files, installing Kata Containers and running E2E tests with that. However, those tests need to be run manually on bare-metal hosts because triple-nested virtualization (like we would have to do in our Azure CI) is too slow.

### CODE REVIEW

**Comment by pohly** [comment]
@devimc: this PR now has testing against and instructions for Kata Containers 1.11.0-rc0. Can you perhaps check that the changes in `docs` look reasonable?

More feedback of course also welcome :grinning: 

I'm pushing this while local tests are still running, but hopefully the new tests work now.

**Comment by pohly** [comment]
kata-deploy fails in our CI (https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/view/pmem-csi/job/pmem-csi/view/change-requests/job/PR-500/29/console):
>  kube-system   pod/kata-deploy-prvzf                              0/1     CrashLoopBackOff   4          4m31s   10.244.1.2 

I'll check tomorrow why it fails there. I worked for me locally.

**Comment by pohly** [comment]
https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/view/pmem-csi/job/pmem-csi/view/change-requests/job/PR-500/lastCompletedBuild/testReport/clear-32690-.lvm-production/E2E/Kata_Containers__Testpattern__Dynamic_PV__ext4___dax_should_support_MAP_SYNC/

> May 11 09:07:37.136: INFO: At 2020-05-11 09:02:31 +0000 UTC - event for dax-volume-test-kata: {kubelet pmem-csi-govm-worker1} FailedCreatePodSandBox: Failed to create pod sandbox: rpc error: code = Unknown desc = container create failed: failed to launch qemu: exit status 1, error messages from qemu log: Could not access KVM kernel module: No such file or directory
qemu-system-x86_64: failed to initialize KVM: No such file or directory

Same error also for other nodes. It looks like we don't have nested virtualization enabled in the Azure VM. Let me see whether I can change that...

File: docs/design.md [hunk 1]
```diff
@@ -126,6 +127,47 @@ In direct device mode, the driver does not attempt to limit space
...
+access to the filesystem prepared by PMEM-CSI is provided inside the
+virtual machine by the 9p or virtio-fs filesystems. Both do not
+support App Direct mode:
+- 9p does not support `mmap` at all.
+- virtio-fs only supports it when not using `MAP_SYNC`, i.e. without dax
+  semantic.
+
+This gets solved as follows:
+- PMEM-CSI creates a volume as usual, either in direct mode or LVM mode.
+- Inside that volume it sets up an ext4 filesystem.
```
Comment (devimc) [inline docs/design.md:143]: jfyi - xfs is also supported

**Comment by pohly** [comment]
Progress (?): without the `-vmx` workaround in `Jenkinsfile` (i.e. plain `-cpu host`), `/dev/kvm` appears in the nodes, but it then fails with "Failed to create pod sandbox: rpc error: code = Unknown desc = container create failed: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing".

**Comment by pohly** [comment]
> "Failed to create pod sandbox: rpc error: code = Unknown desc = container create failed: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing".

@devimc explained on IRC that triple nesting of VMs makes the inner QEMU so slow that the kubelet -> CRI communication times out. This means we cannot test with Kata Containers in the current Azure CI.

I'll make the tests optional. Until we have automatic testing on real hardware (BMaaS!), we'll simply have to test them manually from time to time on real hardware to detect regressions.

File: docs/design.md (hunk 1 above)
Comment (pohly) [reply docs/design.md]: Fixed.

**Comment by pohly** [comment]
Tests are clean now after disabling the Kata Containers tests in our Azure CI.

@avalluri: okay to merge?

File: deploy/common/pmem-storageclass-ext4-kata.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+kind: StorageClass
+metadata:
+  name: pmem-csi-sc-ext4-kata
+parameters:
+  csi.storage.k8s.io/fstype: ext4
+  eraseafter: "true"
+  kataContainers: "true"
+provisioner: pmem-csi.intel.com
+reclaimPolicy: Delete
+volumeBindingMode: Immediate
```
Comment (avalluri) [inline deploy/common/pmem-storageclass-ext4-kata.yaml:11]: As per [app yaml(https://github.com/intel/pmem-csi/pull/500/files#diff-17203e3a5882efb0c1944558564a9c52R10-R11] , the application is expected to run only on nodes with `katacontainers.io/kata-runtime: "true"`. So, if we use immediate binding mode might end up creating a node on the wrong node?

Comment (pohly) [reply deploy/common/pmem-storageclass-ext4-kata.yaml:11]: True. I should better switch this to "late-binding".

File: docs/design.md [hunk 3]
```diff
@@ -126,6 +127,47 @@ In direct device mode, the driver does not attempt to limit space
...
+
+This gets solved as follows:
+- PMEM-CSI creates a volume as usual, either in direct mode or LVM mode.
+- Inside that volume it sets up an ext4 or xfs filesystem.
+- Inside that filesystem it creates a `pmem-csi-vm.img` file that contains
+  partition tables, dax metadata and a partition that takes up most of the
+  space available in the volume.
+- That partition is bound to a `/dev/loop` device and the formatted
+  with the requested filesystem type for the volume.
+- When an applications needs access to the volume, PMEM-CSI mounts
```
Comment (avalluri) [inline docs/design.md:149]: typo: an applications -> an application

**Comment by avalluri** [comment]
@pohly looks good to me. Just go ahead and merge after fixing the storage classes and the documentation nit.

**Pull request merged**